apis/src/main/java/org/apache/rocketmq/apis/producer/ProducerBuilder.java

```diff
@@ -0,0 +1,92 @@
+
+import org.apache.rocketmq.apis.ClientConfiguration;
+import org.apache.rocketmq.apis.retry.BackoffRetryPolicy;
+import org.apache.rocketmq.apis.exception.AuthenticationException;
+import org.apache.rocketmq.apis.exception.AuthorisationException;
+import org.apache.rocketmq.apis.exception.ClientException;
+import org.apache.rocketmq.apis.exception.NetworkConnectionException;
+import org.apache.rocketmq.apis.exception.NetworkTimeoutException;
+import org.apache.rocketmq.apis.exception.TopicDoesNotExistException;
+import org.apache.rocketmq.apis.message.Message;
+
+/**
+ * Builder to config and start {@link Producer}.
+ */
+public interface ProducerBuilder {
+    /**
+     * Set the client configuration for producer.
+     *
+     * @param clientConfiguration client's configuration.
+     * @return the producer builder instance.
+     */
+    ProducerBuilder setClientConfiguration(ClientConfiguration clientConfiguration);
+
+    /**
+     * Declare topics ahead of message sending/preparation.
+     *
+     * <p>Even though the declaration is not essential, we <strong>highly recommend</strong> to declare the topics in
+     * advance, which could help to discover potential mistakes.
+     *
+     * @param topics topics to send/prepare.
+     * @return the producer builder instance.
+     */
+    ProducerBuilder withTopics(String... topics);
+
+    /**
+     * Set the threads count for {@link Producer#sendAsync(Message)}.
+     *
+     * @return the producer builder instance.
+     */
+    ProducerBuilder setAsyncThreadCount(int count);
+
+    /**
+     * Set the retry policy to send message.
+     *
+     * @param retryPolicy policy to re-send message when failure encountered.
+     * @return the producer builder instance.
+     */
+    ProducerBuilder setRetryPolicy(BackoffRetryPolicy retryPolicy);
+
+    /**
+     * Set the transaction checker for producer.
+     *
+     * @param checker transaction checker.
+     * @return the produce builder instance.
+     */
+    ProducerBuilder setTransactionChecker(TransactionChecker checker);
+
+    /**
+     * Finalize the build of {@link Producer} instance and start.
+     *
+     * <p>The producer does a series of preparatory work during startup, which could help to identify more unexpected
+     * error earlier.
+     *
+     * <p>Especially, if this method is invoked more than once, different producer will be created and started.
+     *
+     * @return the producer instance.
+     * @throws TopicDoesNotExistException if there are nonexistent topic(s).
+     * @throws AuthorisationException     if no permission to communicate with server.
+     * @throws AuthenticationException    if identification could not be recognized by server.
+     * @throws NetworkTimeoutException    if encountered network timeout to communicate with server.
+     * @throws NetworkConnectionException if there is a network connection problem.
+     */
+    Producer start() throws ClientException;
```

Comment:
may be move start to the producer.
some case will just build the producer without start.

Reply:
Actually producerBuilder#start could avoid to start producer repeatedly. What's more, a producer which is not started does not make sense.

Reply:
It's strange if we don't have a `build` method in a builder interface.

If we only provide a `start` method in the producer builder, that means all the producers are started by default, so why we should have a `start`?

Reply:
> It's strange if we don't have a `build` method in a builder interface.
> 
> If we only provide a `start` method in the producer builder, that means all the producers are started by default, so why we should have a `start`?

I agree with the opinon that lack of `#build()` may be a little bit wireld, let's look at it another way. Is it unnecessary for the user to start this process? If so, then I think it's just a naming issue.